packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmObjectHelper.kt

```diff
@@ -759,6 +761,141 @@ internal object RealmObjectHelper {
+                                    )
+                                )
+                            }
+                        }
+                    }
+                    PropertyType.RLM_PROPERTY_TYPE_INT -> {
+                        // MutableRealmInt is a special case, since Core treats it as Int
+                        // in the schema. So we need to test for our wrapper class here
+                        val value = accessor.get(source)
+                        if (value is MutableRealmInt) {
+                            accessor.set(target, MutableRealmInt.create(value.get()))
+                        } else {
+                            accessor.set(target, value)
+                        }
+                    }
+                    else -> {
+                        accessor.set(target, accessor.get(source))
+                    }
+                }
+                CollectionType.RLM_COLLECTION_TYPE_LIST -> {
+                    val elements: List<Any?> = accessor.get(source) as List<Any?>
+                    when (property.type) {
+                        PropertyType.RLM_PROPERTY_TYPE_INT,
+                        PropertyType.RLM_PROPERTY_TYPE_BOOL,
+                        PropertyType.RLM_PROPERTY_TYPE_STRING,
+                        PropertyType.RLM_PROPERTY_TYPE_BINARY,
+                        PropertyType.RLM_PROPERTY_TYPE_FLOAT,
+                        PropertyType.RLM_PROPERTY_TYPE_DOUBLE,
+                        PropertyType.RLM_PROPERTY_TYPE_TIMESTAMP,
+                        PropertyType.RLM_PROPERTY_TYPE_OBJECT_ID,
+                        PropertyType.RLM_PROPERTY_TYPE_UUID -> {
+                            accessor.set(target, elements.toRealmList())
+                        }
+                        PropertyType.RLM_PROPERTY_TYPE_OBJECT -> {
+                            val list = UnmanagedRealmList<BaseRealmObject>()
+                            if (currentDepth < maxDepth) {
+                                (elements as List<BaseRealmObject>).forEach { listObject: BaseRealmObject ->
+                                    list.add(
+                                        createDetachedCopy(
+                                            mediator,
+                                            listObject,
+                                            currentDepth + 1,
+                                            maxDepth,
+                                            closeAfterCopy,
+                                            cache
+                                        )
+                                    )
+                                }
+                            }
+                            accessor.set(target, list)
+                        }
+                        else -> {
+                            throw IllegalStateException("Unknown type: ${property.type}")
+                        }
+                    }
+                }
+                CollectionType.RLM_COLLECTION_TYPE_SET -> {
```

Comment:
I thought about this as well, but there are enough differences that I think it is going to just make the code pretty unreadable, and there are no current plans to add more collection types.

Currently, the code differs in 3 locations... 1) converting primitive elements to either list/set 2) creating unmanaged list/set and 3) casting the type. This could be extracted into an interface and delegated to an implementation, but IMO the current approach is more readable, even if it results in a bit of duplicated code.